### PR TITLE
Added tests for cljc files, include-macros and refer-macros. Closes #101.

### LINKED ISSUE #101: Tests for ns macro with require
Add tests for `ns` macro for `include-macros` and `refer-macros` like [here](https://github.com/clojure/clojurescript/wiki/Differences-from-Clojure#namespaces).

Eg.
```
(ns testme.core
  (:require [foo.core :as foo :refer [foo-fn] :include-macros true]
            [woz.core :as woz :refer [woz-fn] :refer-macros [apple jax]]))
```

### PR DESCRIPTION
**Pull request by tomasz-biernacki**
Added tests for possibly every way to require macros. Unfortunately there is still the issue with the order of tests but I will open a separate issue for that. Of course it's not only a problem with tests because it can be easily replicated in the repl. 

Very probably is related to the fact the the `1.7.170` contains bugs in these features, which have been partially solved in `1.7.202`.

So here should be the complete suit of tests.

### CODE REVIEW

File: test/node/replumb/require_node_test.cljs [hunk 1]
```diff
@@ -359,14 +401,87 @@ trim-newline
...
+    ;;   (is (= "36" out) "(f/mul-baz 6 6) should be 36")
+    ;;   (repl/reset-env! '[my.namespace foo.bar.baz]))
+
+    (let [res (do (read-eval-call "(ns my.namespace (:require [foo.bar.core :as f :include-macros true]))")
+                  (read-eval-call "(f/mul-core 30 5)"))
+          out (unwrap-result res)]
+      (is (success? res) "(ns my.namespace (:require...:include-macros...])) and (f/mul-core 30 5) should succeed")
+      (is (valid-eval-result? out) "(ns my.namespace (:require...:as...])) and (f/mul-core 30 5) should be a valid result")
+      (is (= "150" out) "(f/mul-core 30 5) should be 150")
+      (repl/reset-env! '[my.namespace foo.bar.core foo.bar.macros])))
```
Comment (tomasz-biernacki) [inline test/node/replumb/require_node_test.cljs]: The above works in `1.7.170` but not the previous.
In `1.7.202` they both work.

File: dev-resources/private/test/src/cljc/foo/bar/macros.cljc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+(ns foo.bar.macros)
+
+(defmacro str->int [s]
+  #?(:clj  (Integer/parseInt s)
+     :cljs (js/parseInt s)))
```
Comment (arichiardi) [inline dev-resources/private/test/src/cljc/foo/bar/macros.cljc]: Please add a newline or my emacs will go nuts :smile:

File: dev-resources/private/test/src/cljc/foo/bar/self.cljc [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+#?(:clj 
+  (defmacro add
+    [a b]
+    `(+ ~a ~b)))
+
+#?(:cljs
+  (defn sum
+    [a b]
+    (add a b)))
```
Comment (arichiardi) [inline dev-resources/private/test/src/cljc/foo/bar/self.cljc]: here too

File: test/node/replumb/require_node_test.cljs [hunk 4]
```diff
@@ -446,7 +561,15 @@ trim-newline
...
     (ns-macro)
-    (ns-macro-with-require-macro)
+
+    (ns-macro-require-refer-macros)
+    (ns-macro-require-include-macros)
+    (ns-macro-require-macros)
+    (ns-macro-require-macros-refer)
+    (ns-macro-use-macros)
+    (ns-macro-require)
+    (ns-macro-require-macros-as)
```
Comment (tomasz-biernacki) [inline test/node/replumb/require_node_test.cljs]: As said, the order is fixed but the results of the tests are true, I double checked in the repl.
I will open an issue for that.

File: dev-resources/private/test/src/cljc/foo/bar/core.cljc [hunk 5]
```diff
@@ -0,0 +1,16 @@
+(ns foo.bar.core
+  (#?(:clj  :require
+      :cljs :require-macros)
+    [foo.bar.macros]))
```
Comment (arichiardi) [inline dev-resources/private/test/src/cljc/foo/bar/core.cljc]: Why is this in a new line, is it idiomatic like this? To me it should go on the same line as `:require`..

File: dev-resources/private/test/src/cljc/foo/bar/self.cljc [hunk 6]
```diff
@@ -0,0 +1,13 @@
+(ns foo.bar.self
+  #?(:cljs (:require-macros
+      [foo.bar.self :refer [add]])))
```
Comment (arichiardi) [inline dev-resources/private/test/src/cljc/foo/bar/self.cljc]: here too

File: test/node/replumb/require_node_test.cljs [hunk 7]
```diff
@@ -316,8 +318,18 @@ trim-newline
       (is (re-find #"ABCDEF" out) "The result should be ABCDEF")
       (repl/reset-env! '[my.namespace foo.bar.baz])))
 
-  (deftest ns-macro-with-require-macro
-    ;; baz.clj file pared with baz.cljs
+  ;; quux.clj is a single .clj file and namespace
+  ;; baz.clj file is pared with baz.cljs (but with no require-macros in baz.cljs)
```
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: _paired_

File: test/node/replumb/require_node_test.cljs [hunk 8]
```diff
@@ -326,13 +338,34 @@ trim-newline
...
-      (is (valid-eval-result? out) "(ns my.namespace (:require-macros ...)) and (foo.bar.quux/mul-quux 2 2) should be a valid result.")
-      (is (= "4" out) "(foo.quux/mul-quux 2 2) should be 4")
+      (is (success? res) "(ns my.namespace (:require-macros ...)) and (foo.bar.core/mul-core 30 1) should succeed")
+      (is (valid-eval-result? out) "(ns my.namespace (:require-macros ...])) and (foo.bar.core/mul-core 30 1) should be a valid result.")
+      (is (= "30" out) "(foo.bar.core/mul-core 30 1) should be 30")
+      (repl/reset-env! '[my.namespace foo.bar.core foo.bar.macros])))
+
+  (deftest ns-macro-require-macros-as
+    ;; TB - this test fails but shouldn't, see https://github.com/clojure/clojurescript/wiki/Differences-from-Clojure#lisp
+    ;; see also http://dev.clojure.org/jira/browse/CLJS-1449
```
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: Do we have an issue for this? It is easier to track the issue ;)

File: test/node/replumb/require_node_test.cljs [hunk 9]
```diff
@@ -359,14 +401,87 @@ trim-newline
...
+      (repl/reset-env! '[my.namespace foo.bar.core foo.bar.macros])))
+
+  (deftest ns-macro-require
+    ;; cannot require clj file
+    (let [res (read-eval-call "(ns my.namespace (:require [foo.bar.quux]))")
+          error (unwrap-result res)]
+      (is (not (success? res)) "(ns my.namespace (:require [foo.bar.quux])) should not succeed")
+      (is (valid-eval-error? error) "(ns my.namespace (:require [foo.bar.quux])) should be an instance of jsError.")
+      (is (re-find #"No such namespace: foo.bar.quux" (extract-message error))
+          "(ns my.namespace (:require [foo.bar.quux])) should have a valid error message.")
```
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: Let's put this on the same line as `is` even if it's long...

File: test/node/replumb/require_node_test.cljs [hunk 10]
```diff
@@ -359,14 +401,87 @@ trim-newline
...
+
+  (deftest ns-macro-require-include-macros
+    ;; solved in 1.7.202 - see commented test below
+    (let [res (do (read-eval-call "(ns my.namespace (:require [foo.bar.baz :include-macros true]))")
+                  (read-eval-call "(foo.bar.baz/mul-baz 6 6)"))
+          error (unwrap-result res)]
+      (is (not (success? res)) "(ns my.namespace (:require ... :include-macros ...)) and (foo.bar.baz/mul-baz 6 6) should not succeed")
+      (is (valid-eval-error? error) "(ns my.namespace (:require ...:include-macros...)) and (foo.bar.baz/mul-baz 6 6) should be an instance of jsError")
+      (is (re-find #"Cannot read property 'call' of undefined" (extract-message error))
+          "(ns my.namespace (:require ...:include-macros...)) and (foo.bar.baz/mul-baz 6 6) should have correct error message.")
```
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: here too

File: dev-resources/private/test/src/cljc/foo/bar/core.cljc (hunk 5 above)
Comment (tomasz-biernacki) [inline dev-resources/private/test/src/cljc/foo/bar/core.cljc]: Is the same as [here](http://blog.fikesfarm.com/posts/2015-06-19-portable-macro-musing.html)

File: test/node/replumb/require_node_test.cljs [hunk 11]
```diff
@@ -359,14 +401,87 @@ trim-newline
...
+      (repl/reset-env! '[my.namespace foo.bar.baz]))
+
+    ;; (let [res (do (read-eval-call "(ns my.namespace (:require [foo.bar.baz :include-macros true]))")
+    ;;               (read-eval-call "(foo.bar.baz/mul-baz 6 6)"))
+    ;;       out (unwrap-result res)]
+    ;;   (is (success? res) "(ns my.namespace (:require ... :include-macros ...)) and (f/mul-baz 6 6) should succeed")
+    ;;   (is (valid-eval-result? out) "(ns my.namespace (:require ...:include-macros...)) and (f/mul-baz 6 6) should be a valid result.")
+    ;;   (is (= "36" out) "(f/mul-baz 6 6) should be 36")
+    ;;   (repl/reset-env! '[my.namespace foo.bar.baz]))
```
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: Why is this commented out?

Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: Apologies, I read above :+1:

File: dev-resources/private/test/src/cljc/foo/bar/core.cljc (hunk 5 above)
Comment (arichiardi) [inline dev-resources/private/test/src/cljc/foo/bar/core.cljc]: What do you prefer? I vote for my option (of course) but it's a personal preference I guess..

Comment (tomasz-biernacki) [inline dev-resources/private/test/src/cljc/foo/bar/core.cljc]: For me is ok, everything one the same line?
`(#?(:clj  :require :cljs :require-macros) [foo.bar.macros]))`
or
```
(#?(:clj  :require
     :cljs :require-macros) [foo.bar.macros]))
```

File: test/node/replumb/require_node_test.cljs (hunk 4 above)
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: Too bad that negating tests is not that easy I wish I could do:

```
(defn test-ns-hook []
  
  (ns-macro-require-macros)
  (ns-macro-require-include-macros)
  (not (ns-macro-require-macros-refer)) 
```

Other order, but all the tests become `is-not`...

File: test/node/replumb/require_node_test.cljs (hunk 1 above)
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: I would open an issue with a link to the comment test + comment in code

File: dev-resources/private/test/src/cljc/foo/bar/core.cljc (hunk 5 above)
Comment (arichiardi) [inline dev-resources/private/test/src/cljc/foo/bar/core.cljc]: Maybe second

File: test/node/replumb/require_node_test.cljs (hunk 8 above)
Comment (arichiardi) [inline test/node/replumb/require_node_test.cljs]: TB?

**Pull request merged**